query/src/main/java/org/infinispan/query/affinity/AffinityIndexManager.java

```diff
@@ -0,0 +1,286 @@
+         logDebug("Ownership lost, closing index manager");
+         nextOwner = newOwner;
+         flushAndReleaseResources();
+         hasOwnership = false;
+      } finally {
+         writeLock.unlock();
+      }
+   }
+
+   private void handleOwnershipAcquired() {
+      writeLock.lock();
+      try {
+         logDebug("Ownership acquired");
+         nextOwner = null;
+         hasOwnership = true;
+      } finally {
+         writeLock.unlock();
+      }
+   }
+
+   @Override
+   public void flushAndReleaseResources() {
+      Address lockHolder = getLockHolder(getIndexName());
+      if (lockHolder != null) {
+         InfinispanDirectoryProvider directoryProvider = (InfinispanDirectoryProvider) getDirectoryProvider();
+         int activeDeleteTasks = directoryProvider.getActiveDeleteTasks();
+         while (activeDeleteTasks > 0) {
+            try {
+               Thread.sleep(POLL_WAIT);
+               logDebug("Waiting for pending delete tasks, remaining: %s", activeDeleteTasks);
+               activeDeleteTasks = directoryProvider.getActiveDeleteTasks();
+            } catch (InterruptedException e) {
+               Thread.currentThread().interrupt();
+            }
+         }
+         logDebug("Flushing directory provider");
+         super.flushAndReleaseResources();
+      }
+   }
+
+   @Override
+   public void performStreamOperation(LuceneWork singleOperation, IndexingMonitor monitor, boolean forceAsync) {
+      super.performStreamOperation(singleOperation, monitor, forceAsync);
+   }
+
+   private Object stringToKey(String key) {
+      return keyTransformationHandler.stringToKey(key, cache.getAdvancedCache().getClassLoader());
+   }
+
+   private int getSegment(Object key) {
+      return distributionManager.getConsistentHash().getSegment(key);
+   }
+
+   private Address getLockHolder(String indexName) {
+      InfinispanDirectoryProvider directoryProvider = (InfinispanDirectoryProvider) getDirectoryProvider();
+      return directoryProvider.getLockOwner(indexName, IndexWriter.WRITE_LOCK_NAME);
+   }
+
+   private Address getLocation(LuceneWork work) {
+      Address localAddress = rpcManager.getAddress();
+      if (work.getIdInString() == null) {
+         return localAddress;
+      }
+      int segment = getSegment(stringToKey(work.getIdInString()));
+      Address lockHolder = getLockHolder(replaceShard(getIndexName(), segment));
+      logDebug("Lock holder for %d is %s", segment, lockHolder);
+      Address destination;
+      if (!hasOwnership) {
+         logDebug("Lost ownership, new owner is %s", nextOwner);
+         destination = nextOwner;
+      } else {
+         destination = (lockHolder != null && !localAddress.equals(lockHolder)) ? lockHolder : localAddress;
+      }
+      return destination;
+   }
+
+   private Map<Address, List<LuceneWork>> partitionWorkByAddress(List<LuceneWork> works) {
+      return works.stream().collect(Collectors.toMap(this::getLocation, Arrays::asList, (w1, w2) -> {
```

Comment:
But since we are here...If there's a key collision, the merge function will fail with ```UnsupportedOperationException``` because ```Arrays.asList(T...)``` produces a list that cannot be resized, will have it sorted out